File: Codes/010_isMatch/10.正则表达式匹配.go
/*
 * @lc app=leetcode.cn id=10 lang=golang
 *
 * [10] 正则表达式匹配
 */

// @lc code=start
func match(s string, p string, i int, j int) bool {
	if i == 0 {
		return false
	} else if p[j-1] == '.' {
		return true
	} else {
		return s[i-1] == p[j-1]
	}
}

func isMatch(s string, p string) bool {
	//1. re 作弊法
	// r, _ := regexp.MatchString("^"+p+"$", s)
	// return r
	//2 动态规划法
	// 存储数据
	var ds [][]bool

	for i := 0; i < (len(s) + 1); i++ {
		ds = append(ds, make([]bool, len(p)+1))

		ds[i][0] = false
		if i == 0 {
			ds[0][0] = true
		}
		for j := 1; j < (len(p) + 1); j++ {
			if p[j-1] == '*' {
				ds[i][j] = ds[i][j-2]
				if match(s, p, i, j-1) {
					ds[i][j] = ds[i-1][j] || ds[i][j]
				}
			} else {
				if match(s, p, i, j) {
					ds[i][j] = ds[i-1][j-1]
				} else {
					ds[i][j] = false
				}
			}
		}
	}

	return ds[len(s)][len(p)]

}

// @lc code=end

File: Codes/010_isMatch/10.正则表达式匹配.js
/*
 * @lc app=leetcode.cn id=10 lang=javascript
 *
 * [10] 正则表达式匹配
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    // 1. 作弊法
    // var n=s.search('^'+p+'$')
    // if(n!=-1){
    //     return true
    // }else{
    //     return false
    // }

    // 2. 动态规划法
    var match=function(i,j){
        if(i==0){
            return false
        }else if(p[j-1]=="." || p[j-1]==s[i-1]){
            return true
        }else{
            return false
        }
    }

    //信息存储数组
    var ds=new Array();

    // i必须从0开始，因为可能有空匹配的情况 s='' p="a*b*c*"
    for(i=0;i<s.length+1;i++){
        
        ds[i]=new Array();
        if(i==0){
            ds[0][0]=true
        }else{
            ds[i][0]=false
        }

        for(j=1;j<p.length+1;j++){
            
            if(p[j-1]=="*"){
                //情况1 不匹配 例 p: ab*c=>ac  s：abc
                ds[i][j]=ds[i][j-2]
                
                //情况2 匹配   例 p: ab*c      s：abc=>ac  
                if(match(i,j-1)){
                    ds[i][j]=(ds[i-1][j]||ds[i][j])
                }
            }else{
                // 普通匹配了 例 p: ab*c=>b*c  s: abc=>b*c
                if(match(i,j)){
                    ds[i][j]=ds[i-1][j-1]
                }else{
                    ds[i][j]=false
                }
            }
        }
    }
    return ds[s.length][p.length]
};
// @lc code=end

